Add adc support to php client lib

ajs424 opened the pull request:
add ADC support

bshaffer commented on examples/Authentication/google_ads_php.ini:
```diff
@@ -23,6 +23,13 @@ developerToken = "INSERT_DEVELOPER_TOKEN_HERE"
+; `gcloud auth application-default login`. ADC takes precedence over other OAuth2 settings
+; in this section if enabled.
+; See https://cloud.google.com/docs/authentication/application-default-credentials for details.
+; useApplicationDefaultCredentials = true
```
Hmm, so in this case, the "default" would override the explicitly set `clientId`/`clientSecret`/`refreshToken` values below. I would suggest that the behavior should be the application default credentials kick in when the values below aren't found. If this can be done in a backwards-compatible way, we should just make that happen all the time (and not require a parameter set to true)

bshaffer commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -156,23 +157,48 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+                $credentials = ApplicationDefaultCredentials::getCredentials($scopes);
+                error_log("Successfully obtained credentials via Application Default Credentials.");
+                return $credentials; // Return successfully found ADC credentials.
+            } catch (\Exception $e) {
```
we should use `DomainException` here, as that's what this method has defined in `@throws`
```suggestion
            } catch (\DomainException $e) {
```

ajs424 commented:
@AnashOommen @Raibaz 

updated changes: ADC should be an implicit fallback that automatically activates when no explicit credentials are provided, thus eliminating the need for the useApplicationDefaultCredentials flag. The other key point is to use appropriate exceptions like DomainException for credential failures.

Raibaz commented on examples/Authentication/google_ads_php.ini:
```diff
@@ -23,6 +23,15 @@ developerToken = "INSERT_DEVELOPER_TOKEN_HERE"
+; environment variable, 'gcloud auth application-default login', or compute engine service account).
+; Explicitly set credentials take precedence over ADC.
+; See https://cloud.google.com/docs/authentication/application-default-credentials for details.
+; The 'useApplicationDefaultCredentials' flag has been removed, as ADC is now the default fallback.
```
Did we ever release a version of the client library with the `useApplicationDefaultCredentials` flag?

If we didn't, we can just remove this line.

fiboknacky commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -18,11 +18,16 @@
 
+use DomainException;
+use Google\Auth\ApplicationDefaultCredentials; 
+//use Google\Auth\CredentialsLoaderException;
```
Remove?

fiboknacky commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -18,11 +18,16 @@
 use InvalidArgumentException;
 use UnexpectedValueException;
+use Google\Ads\GoogleAds\Util\EnvironmentalVariables;
+// ...
```
Remove?

fiboknacky commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -156,24 +171,56 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+                ]
+            );
+        }
+      // 3. FALLBACK: Use Application Default Credentials (ADC)
```
Wrong indentation.

fiboknacky commented on tests/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilderTest.php:
```diff
@@ -1,47 +1,22 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
```
This shouldn't be removed.

AnashOommen commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -155,24 +169,64 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+    public function build(): FetchAuthTokenInterface
+    {
+        // The default scope for Google Ads API.
+        $adsScope = ['https://www.googleapis.com/auth/adwords'];
```
Is this configurable? The scopes are configurable in some other libraries, so maybe this should come from a config variable?

fiboknacky replied:
I added a comment above actually but the code was outdated:

```
I'm not 100% sure about the change here, but we do have object properties $this->scopes too.
Shouldn't we use that?
And if this variable is always used once here, why don't we just hard-code it below or declare a constant above in this class like other classes?
```

fiboknacky replied:
Added more comments.

fiboknacky reviewed (commented):
Added a bunch of comments.

ajs424 commented:
Hi @fiboknacky 

Hi [Reviewer Name, e.g., @john_doe],

I went ahead and worked on a refactor implemented all the requested changes. All tests are currently passing. I believe PR should be good to merge but please take a look.

Key updates based on feedback:

Scope Configuration: Removed the redundant local variable. Scopes for the User Refresh and ADC fallback flows are now configurable using the $this->scopes property, with the default scope applied via defaultOptionals(), as suggested.

Builder Pattern Fix: Removed adcFetcher from the constructor and added the test-only setter to maintain adherence to the Builder design pattern.

Service Account Validation: Updated the Service Account flow validation to correctly check if the scope is the generic default

fiboknacky commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -155,23 +170,65 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
-    public function build()
+    public function build(): FetchAuthTokenInterface
     {
-        $this->defaultOptionals();
```
If you remove the call here, where this defaultOptionals() are called then?

fiboknacky commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -155,23 +170,65 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+                );
+            }
+            // Service Account flow uses the specific configured scope string
+            $scopesForExplicitFlows = explode(' ', $this->scopes);
```
This and `$scopeArrayForUserAndAdc` look the same. You probably don't need so many variables.

ajs424 commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -155,23 +170,65 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+                    "Both 'jsonKeyFilePath' and 'scopes' must be set when using service account flow."
+                );
+            }
+            if (!is_null($this->clientId) || !is_null($this->clientSecret) || !is_null($this->refreshToken)) {
```
So it looks like its technically 119 lines :) but agreed - probably can be improved from a readability standpoint so I'll go ahead and wrap it

ajs424 commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -252,17 +260,17 @@ public function validate()
-            || is_null($this->refreshToken)
+            !is_null($this->clientId)
+            || !is_null($this->clientSecret)
+            || !is_null($this->refreshToken)
```
The structure is designed to handle the ADC fallback state.

The outer elseif acts as a Flow Detector. Its job is to check if the user has started the Installed/Web App Flow by setting any of the three required fields.

If the flow is started (Detector is true), the inner if throws an exception only if the flow is incomplete (i.e., one or more fields are missing).

If we didn't have the outer detector, the validation would incorrectly throw an error when the user provides no credentials at all, which is the required state for the ADC fallback.

I will add a comment to clarify

fiboknacky commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -150,28 +165,84 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+        try {
+            // Use the determined scope array, allowing configuration via $this->scopes
+            return call_user_func($this->adcFetcher, $this->scopes);
+        } catch (\Google\Auth\CredentialsLoaderException $e) {
```
Nit: You can import this class at the beginning of the file.

ajs424 commented on src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php:
```diff
@@ -150,28 +165,84 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
+    public function build(): FetchAuthTokenInterface
     {
         $this->defaultOptionals();
-        $this->validate();
```
Yes, it is intentional. The call to $this->validate() was removed from the build() method in favor of placing the validation logic inline within build()
